fix: prevent infinite loop on prettyDOM calls

tsirlucas opened the pull request:
### Description

Fixed infinite loop described in ~https://github.com/vitest-dev/vitest/pull/7152~ https://github.com/vitest-dev/vitest/issues/7249

Thought about adding a test case but it would just enter the loop and timeout the pipeline instead of failing (this was happening to my setup in the private repo haha). WDYT?

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->

<!-- You can also add additional context here -->

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [x] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

tsirlucas commented on packages/utils/src/display.ts:
```diff
@@ -72,9 +72,11 @@ export function stringify(
-
+  
+  // Prevents infinite loop https://github.com/vitest-dev/vitest/issues/7249
+  const nextMaxDepth = maxDepth === Infinity ? Number.MAX_VALUE : Math.floor(maxDepth / 2)
```
This is stil a pretty big recursion so I wonder how it will behave in CI env with limited resources. Will run a test and get back here.

tsirlucas replied:
Ok I have pretty similar results time-wise doesnt matter if I use `Number.MAX_VALUE` or just `10`.

hi-ogawa commented:
I think we can avoid infinity for `prettyDOM`. I'd imagine practically it wouldn't matter whether the dom is truncated at 100, or even just 10. 

https://github.com/vitest-dev/vitest/blob/f9a628438a5462436b59dd9bdeffddada19a9e81/packages/browser/src/client/tester/public-utils.ts#L58

Pretty-format's maxDepth is quite scary and, for example, we reduced from 10 to 8 in other cases https://github.com/vitest-dev/vitest/pull/7133

Can you add test cases with repro from https://github.com/vitest-dev/vitest/issues/7249#issuecomment-2591651590?

tsirlucas commented on packages/browser/src/node/providers/playwright.ts:
```diff
@@ -185,6 +185,12 @@ export class PlaywrightBrowserProvider implements BrowserProvider {
       })
     }
 
+    // unhandled page crashes will hang vitest process
```
this was making the tests hang https://github.com/vitest-dev/vitest/issues/7249#issuecomment-2592373055

tsirlucas commented on packages/utils/src/display.ts:
```diff
@@ -73,8 +73,10 @@ export function stringify(
   }
 
+  // Prevents infinite loop https://github.com/vitest-dev/vitest/issues/7249
+  const nextMaxDepth = maxDepth === Infinity ? Number.MAX_SAFE_INTEGER : Math.floor(maxDepth / 2)
```
MAX_NUMBER would still crash the browser, so Im using MAX_SAFE_INTEGER which works fine. If you want me to use any specific number, please send it here so I can replace

tsirlucas commented on test/browser/test/browser-crash.test.ts:
```diff
@@ -0,0 +1,15 @@
+import { it } from 'vitest';
+
+it.skip('fails gracefully when browser crashes', async () => {
```
I dont know what to do with this test. It reproduces the issue where browser crashes, but if I enable it it will break the pipeline

sheremet-va replied:
You can add this test to fixtures and create a new test file in test/browser/specs to run the fixture and assert that it fails correctly

tsirlucas commented:
ok i think im hitting a wall with the window tests... any idea what it may be?

tsirlucas commented:
ok... the old 20k nodes test was messing up with perf/stability in CI.

i adjusted the test case to test a simpler scenario that would still break without the changes in PR

also adjusted how we simulate crashes for same reason as above.

tsirlucas commented on packages/utils/src/display.ts:
```diff
@@ -73,8 +73,11 @@ export function stringify(
+  const nextMaxDepth = maxDepth > MAX_DEPTH ? MAX_DEPTH : Math.floor(maxDepth / 2)
   return result.length >= MAX_LENGTH && maxDepth > 1
-    ? stringify(object, Math.floor(maxDepth / 2))
+    ? stringify(object, nextMaxDepth, { maxLength, ...options })
```
i noticed that maxLength and options were not being sent to recursive child calls before so i added it here

hi-ogawa replied:
Maybe one liner?

```suggestion
    ? stringify(object, Math.floor(Math.min(maxDepth, Number.MAX_SAFE_INTEGER) / 2), { maxLength, ...options })
```

sheremet-va reviewed (approved):
LGTM!
